src/hardware/pcspeaker.cpp

```diff
@@ -301,56 +341,34 @@ static void PCSPEAKER_CallBack(Bitu len) {
 		}
 		*stream++=(Bit16s)(value/sample_add);
 	}
-	if(spkr.chan) spkr.chan->AddSamples_m16(len,(Bit16s*)MixTemp);
-
-	//Turn off speaker after 10 seconds of idle or one second idle when in off mode
-	bool turnoff = false;
-	Bitu test_ticks = PIC_Ticks;
-	if ((spkr.last_ticks + 10000) < test_ticks) turnoff = true;
-	if((spkr.mode == SPKR_OFF) && ((spkr.last_ticks + 1000) < test_ticks)) turnoff = true;
-
-	if(turnoff){
-		if(spkr.volwant == 0) { 
-			spkr.last_ticks = 0;
-			if(spkr.chan) spkr.chan->Enable(false);
-		} else {
-			if(spkr.volwant > 0) spkr.volwant--; else spkr.volwant++;
-		
-		}
-	} 
-
```

Comment:
I have a concern (perhaps unjustified) about removal of this code - previously this calculation here was time-based, and it got replaced by calculation based on a number of callbacks received? Is this actually ok, or should FadeVolume use time difference for calculation?

Reply:
Both the old and new shutdown functions reset themselves when a game generates new frequencies.

The previous code does this with book keeping  and comparing the tick timer, which increments in 1ms on every callback.  As the number of idle passes grows, it keeps adding to the last tick.  But the last tick gets reset to zero in the pwm and pit calls to the speaker.

The new code also inspects the state of the speaker every 1ms for each callback of the channel (so happens at the same pacing).  However, instead of the book keeping with ticks, we simply look at the 'pos' value to see if new frequencies have been added, which avoids all the tick book keeping. .

So the trigger to ramp down both occur every 1ms, pass through the callback.

However, we ramp down sooner, and with a mich steeper drop in volume over a shorter period. (Typically audio engineers permit volume ramps in less than 2ms, which is inaudible).. versus over a 10s in the original code. 

Many game 'hit' the speaker once on startup, even if they have no plans of using it (say, using GUS or SB).  The earlier idle detection and fast rampdown lets us scrub that idle DC offset very quickly before it has a chance to impact other audio channels.

You can see that in jungle strike.. it hits the speaker right away, and then goes idle.  We detect and get rid of the offeset before it's intro audio kicks in.

Reply:
Regarding this particular piece of code, this may be of interest:
https://github.com/joncampbell123/dosbox-x/issues/1530
https://github.com/joncampbell123/dosbox-x/commit/068e92788c967d4008a92ebda75914683baea489#diff-fe75d7a776a5217b68926089a61d92ad

Reply:
@rderooy , thanks for the reminder! I remember seeing this engine sound issue on Vogons, but I forgot to test this PR against it.

With this PR, the Sopwith engine sound is sustained indefinitely (press 'S' for single player, then 'S' again to turn-on audio).  Here's the first 50-seconds sitting on the tarmac:

![Screenshot at 2020-06-16 23-22-22](https://user-images.githubusercontent.com/1557255/84863555-edeb7400-b029-11ea-814f-e15ed519867a.png)

Near the end, I accelerate and immediately crash :-)  After I crash, there is no audio and you can see the this PR's DC-zero'ing behavior kick in quickly (up-close):

![Screenshot at 2020-06-16 23-23-15](https://user-images.githubusercontent.com/1557255/84863541-e88e2980-b029-11ea-8179-08e72876d266.png)

This PR discards with the indirect last-tick book-keeping and instead we inspect the `pos` value directly.  The `pos` value is always non-zero whenever the speaker is requested to make a frequency (via the PWM or PIT registers). 

`pos` is decremented as the sound is generated until it hits zero when there is no waveform left to generated (at which point, this PR fade-counter begins to pace downward each `pos`-less callback loop).

Reply:
Paratrooper is another early DOS game that has issues with PC Speaker in vanilla DOSBox.

https://github.com/joncampbell123/dosbox-x/issues/1492

Reply:
> Paratrooper is another early DOS game that has issues with PC Speaker in vanilla DOSBox.
> 
> [joncampbell123/dosbox-x#1492](https://github.com/joncampbell123/dosbox-x/issues/1492)

You're a goldmine of info @rderooy ; thanks for the tip! Will check out Paratrooper now.

Reply:
Yes; this PR does not add new behavior in `timer.cpp` - so it is still affected. 

(This PR only tackles the DC-offset introduced by the PC-Speaker into the mixer, which then drags the other audio device channels positive or negative - which can cause clipping and other nasty effects).

The fix you've lined for paratrooper looks small and specific - perhaps we can add that in another PR; although given it touches the timer device, I'm unsure how wide of an effect this change would have.